node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/HostFlavorUpgrader.java

```diff
@@ -0,0 +1,104 @@
+ * cancelled, and upgrade is attempted of the next host on an old flavor, if any.
+ *
+ * Once a host has been marked for upgrade, {@link HostResumeProvisioner} will complete provisioning of the replacement
+ * host.
+ *
+ * @author mpolden
+ */
+public class HostFlavorUpgrader extends NodeRepositoryMaintainer {
+
+    private final HostProvisioner hostProvisioner;
+    private final Random random;
+    private final Deployer deployer;
+    private final Metric metric;
+
+    public HostFlavorUpgrader(NodeRepository nodeRepository, Duration interval, Metric metric, Deployer deployer, HostProvisioner hostProvisioner) {
+        super(nodeRepository, interval, metric);
+        this.hostProvisioner = Objects.requireNonNull(hostProvisioner);
+        this.deployer = Objects.requireNonNull(deployer);
+        this.metric = Objects.requireNonNull(metric);
+        this.random = new Random(nodeRepository.clock().millis());
+    }
+
+    @Override
+    protected double maintain() {
+        if (!nodeRepository().zone().cloud().dynamicProvisioning()) return 1.0; // Not relevant in zones with static capacity
+        if (nodeRepository().zone().environment().isTest()) return 1.0; // Short-lived deployments
+        if (!nodeRepository().nodes().isWorking()) return 0.0;
+
+        NodeList allNodes = nodeRepository().nodes().list();
+        if (!NodeMover.zoneIsStable(allNodes)) return 1.0;
+        return upgradeHostFlavor(allNodes);
+    }
+
+    private double upgradeHostFlavor(NodeList allNodes) {
+        NodeList activeNodes = allNodes.nodeType(NodeType.tenant)
+                                       .state(Node.State.active)
+                                       .shuffle(random); // Shuffle to avoid getting stuck trying to upgrade the same host
+        int attempts = 0;
+        int failures = 0;
+        for (var node : activeNodes) {
+            Optional<Node> parent = allNodes.parentOf(node);
+            if (parent.isEmpty()) continue;
+            if (!hostProvisioner.canUpgradeFlavor(parent.get(), node)) continue;
+            if (parent.get().status().wantToUpgradeFlavor()) continue; // Already upgrading
+
+            boolean redeployed = false;
+            boolean deploymentValid = false;
+            try (MaintenanceDeployment deployment = new MaintenanceDeployment(node.allocation().get().owner(), deployer, metric, nodeRepository(), true)) {
+                deploymentValid = deployment.isValid();
+                if (!deploymentValid) continue;
+
+                attempts++;
+                log.log(Level.INFO, () -> "Redeploying " + node.allocation().get().owner() + " to upgrade flavor (" +
+                                          parent.get().flavor().name() + ") of " + parent.get());
+                upgradeFlavor(parent.get(), true);
+                deployment.activate();
+                redeployed = true;
+                return asSuccessFactorDeviation(attempts, failures);
```

Comment:
Is it really a failure that no capacity of the newer flavor is available at the moment? If we assume that is quite normal (i.e not something ops should look into) I think it would be better to return 1 here.

Reply:
Good point, will change it.